### ISSUE TITLE
Add option to validate certificate trust against rDNS results

### ISSUE THREAD

**Issue by fishouttah20** [drwetter/testssl.sh#1116]
Please make sure that you provide enough information so that we understand what your issue is about.
When scanning based on IP address the Certificate Trust (hostname) will always say "certificate does not match supplied URI". This is a feature request that where the IP resolves via rDNS to a domain name, that rDNS name is checked against the SAN of the Certificate and if they match, the does not match URI error is not triggered.
   
1. testssl version from the banner (testssl.sh -b 2>/dev/null | head -4 | tail -2)
3.0rc1 (9a3b6e3 2018-09-03 13:25:19 -- )
2. what exactly was happening, output is needed
When parallel scanning a large number of IP addresses all of them result in Trust (hostname) errors of Certificate does not match supplied URI
3. what did you expect instead?
I didn't really expect it, but it would be nice to have it validate hostname against the rDNS domain (where applicable)
4. steps to reproduce

   1. testssl.sh command line
 ./testssl.sh --file nmapscan.gnmap --parallel -p -S -P --openssl-timeout 60 -oA results-file
    1. if possible: target IP
 DM if needed
    1. openssl version used (testssl.sh -b 2>/dev/null | head -16 | tail -3)
 1.0.2-chacha (1.0.2i-dev)
    1. your operating system (uname -a)
Ubuntu 4.15.0-29-generic x86_64

**Comment by fishouttah20** [comment]
It looks like I might have jumped the gun here as you're sort of already doing this. When using --file with the output from nmap, it looks like you're parsing the results into a txt file and when nmap supplies an rDNS name, it populates that in the text file to use instead of the IP address. This works well for little shops where DNS always resolves to the same IP, but when load balancing across multiple data centres, having testssl.sh scan the domain name won't always result in the expected IP address results being scanned.

The --ip option will override DNS and hit a specific IP, but that doesn't help when mass scanning. You'd need to parse gnmap in a way that the IP address made it into the .txt file and got passed into the specific instance command-line generated by --parallel

**Comment by drwetter** [comment]
yes, I remember doing this, at least this is how it is supposed to work.

> This works well for little shops where DNS always resolves to the same IP, but when load balancing across multiple data centres, having testssl.sh scan the domain name won't always result in the expected IP address results being scanned.

You got to help me here. E.g. ``mydomain.com`` resolves to ``10.10.10.1`` and ``10.10.10.2``. You scanned /24 of 10.10.10.10. nmap then inserts for the IP addresses above the rDNS value ``cloudaddress-1.domain.tld`` and ``cloudaddress-2.domain.tld``.  How is nmap or testssl.sh supposed to draw a conclusion just from DNS that those IPs belong to ``mydomain.com``?

OT: One could use the certificate's CN/SAN of the default virtual host and feed it into testssl.sh and then start a scan with that DNS. However I think I don't want to to this automatically as this is more a topic for DNS reconnaissance tools. And testssl.sh should be still a tool like a tool in your hand, not smarter than the user.

**Comment by fishouttah20** [comment]
I think you're imagine a scenario where people use rDNS properly to identify specific host names. Imagine a strange world where people have rDNS records on multiple IPs that both point to mydomain.com.

So mydomain.com resolves (note when I say resolves, I don't mean there are two A records, I mean a global traffic manager where it randomly responds with one of a pool of IPs) to 10.10.10.1 in DC1 and 10.10.20.1 in DC2. 

nmap does a scan of both the 10.10.10.0/24 and 10.10.20.0/24 and automatically adds rDNS information into the gnmap file but the IP addresses are still there. testssl.sh, rather than scanning the IP addresses in the gnmap file, scans the domain name mydomain.com and the IP address that's actually scanned is random based on whichever IP address DNS responded with, rather than every IP with open HTTPS ports in both data centres as originally intended with the nmap scan.

Or imagine I'm only scanning one data centre, 10.10.10.0/24 but any of the domains with rDNS for a load balanced domain will actually result in randomly some of those hosts being scanned in the 10.10.20.0/24 range of the second DC.

Re: OT, yeah agreed, testssl.sh doesn't need to be a reconnaissance tool.
